HomeWork Java Unit-тестирование v1

AKnazzz opened the pull request:
ADD:

- [x] Покрыт unit-тестами класс LibraryManager;
- [x] 100% покрытия методов и строк;
- [x] Именование тестов через @DisplayName);
- [x] message в тестах для уточнения ожидаемого результата; 
- [x] Написан параметризованный тест;
- [x] Добавлены комментариях к тестам предложения по улучшению логики.

AKnazzz committed:
Выполнение домашней работы в ветке dev.

ADD:
- Покрыт unit-тестами класс LibraryManager;
- 100% покрытия методов и строк;
- Написан параметризованный тест;
- Добавлены комментариях к тестам предложения по улучшению логики.

blackoutks commented on src/test/java/ru/hh/school/unittesting/homework/LibraryManagerTest.java:
```diff
@@ -0,0 +1,313 @@
+        String userId1 = "user1";
+        String userId2 = "user2";
+
+        libraryManager.borrowBook(bookId, userId2);
```
Все бы хорошо, но тут `borrowBook` вернет `false`, т.к. `user2` будет считаться неактивным. В итоге, книга не будет взята.

AKnazzz replied:
Добавил:
---> поведение мока для активации userId2;
---> также не хватало добавление доступной копии book1 (добавил для корректности).
===> В итоге решился вопрос с проверкой последнего ветвления (branch) - теперь 100% везде =)

blackoutks commented on src/test/java/ru/hh/school/unittesting/homework/LibraryManagerTest.java:
```diff
@@ -0,0 +1,313 @@
+
+    @Test
+    @DisplayName("returnBook тест позитивный: изменение количества книг в инвентаре при возврате книги.")
+    void testReturnBookIncreasesInventory() {
```
Как будто бы этот тест уже избыточен. Можно проверку `getAvailableCopies` добавить во все предыдущие, которые проверяют `returnBook`, а этот тест убрать.

blackoutks commented on src/test/java/ru/hh/school/unittesting/homework/LibraryManagerTest.java:
```diff
@@ -0,0 +1,313 @@
+        libraryManager.addBook("book3", 10);
+        assertEquals(10, libraryManager.getAvailableCopies("book3"),
+                "Ожидается что метод вернёт 10 экземпляров книги book3, книга в наличии  10 шт.");
+        assertEquals(0, libraryManager.getAvailableCopies("book2"),
```
Эту проверку стоит вынести в отдельный тест.

blackoutks commented on src/test/java/ru/hh/school/unittesting/homework/LibraryManagerTest.java:
```diff
@@ -0,0 +1,313 @@
+        libraryManager.addBook("book5", 0);
+        assertEquals(-5, libraryManager.getAvailableCopies("book4"),
+                "Ожидается что метод вернёт `-5` экземпляров книги book4, кол-во книги при добавлении `-5`");
+        assertEquals(0, libraryManager.getAvailableCopies("book5"),
```
Тоже стоило бы разнести по разным тестам. Либо убрать весь метод вообще. Ну или, как вариант, сделать один параметризованный (вероятно - самый правильный вариант, если хочется сохранить все эти проверяемые кейсы). Много тестов - не всегда хорошо :)

AKnazzz replied:
Убрал чрезмерные проверки с сохранением общей проверки логики

blackoutks commented on src/test/java/ru/hh/school/unittesting/homework/LibraryManagerTest.java:
```diff
@@ -0,0 +1,313 @@
+        });
+
+        assertThrows(IllegalArgumentException.class, () -> {
+            libraryManager.calculateDynamicLateFee(-10, false, false);
```
Метод проверяется с почти одинаковыми параметрами два раза подряд в одном тесте. Не очень понятно, что это дает. Думаю, что вторую проверку можно спокойно убрать и даже не выносить в отдельный метод, т.к. это не выглядит как какой-нибудь корнер кейс.

AKnazzz replied:
Убрал дублирующую проверку

blackoutks commented on src/test/java/ru/hh/school/unittesting/homework/LibraryManagerTest.java:
```diff
@@ -0,0 +1,313 @@
+
+    @Test
+    @DisplayName("calculateDynamicLateFee тест позитивный: книга-бестселлер без скидок")
+    void testCalculateDynamicLateFeeBestseller() {
```
Этот тест и два следующих уже учтены в параметризованном выше. Можно удалять.

blackoutks commented on src/test/java/ru/hh/school/unittesting/homework/LibraryManagerTest.java:
```diff
@@ -0,0 +1,313 @@
+
+        libraryManager.addBook(bookId, -3);
+        boolean resultSecond = libraryManager.borrowBook(bookId, userId);
+        assertFalse(resultSecond, "Метод должен вернуть false, если кол-во книги отрицательное.");
```
Можно было бы разнести по разным тестам, т.к. выше ты уже проверяешь то, как работает вызов `addBook` несколько раз подряд. Одна проверка тут выглядела бы проще и понятнее.

blackoutks reviewed (approved):
Немного перестарался с тестами :)

AKnazzz committed:
FIX:
- Убрал излишний @BeforeEach (использую только @InjectMocks).
- Убрал чрезмерные проверки согласно комментариям, упростил код.
- Разбелил тесты для соблюдения принципов 1 тест = 1 проверка.
- Покрыл тестами все возможные варианты развилок (без рефлексии) - сейчас покрытие кода 100%.
